File: 104-maximum-depth-of-binary-tree/solution.cpp
#include <deque>
#include <tuple>

struct TreeNode
{
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
  int maxDepth(TreeNode *root)
  {
    if (root == nullptr)
    {
      return 0;
    }

    std::deque<std::tuple<int, const TreeNode *>> work_stack{{1, root}};
    int max_depth = 1;
    while (!work_stack.empty())
    {
      const auto depth_curr = work_stack.back();
      work_stack.pop_back();
      const int depth = std::get<0>(depth_curr);
      const TreeNode *curr = std::get<1>(depth_curr);
      max_depth = std::max(max_depth, depth);

      if (curr->left != nullptr)
      {
        work_stack.emplace_back(depth + 1, curr->left);
      }
      if (curr->right != nullptr)
      {
        work_stack.emplace_back(depth + 1, curr->right);
      }
    }

    return max_depth;
  }
};

File: 104-maximum-depth-of-binary-tree/solution.py
from collections import deque
from typing import Optional


class TreeNode:
  def __init__(self, val=0, left: Optional['TreeNode'] = None, right: Optional['TreeNode'] = None):
    self.val = val
    self.left = left
    self.right = right


class Solution:
  def maxDepth(self, root: Optional[TreeNode]) -> int:
    if root is None:
      return 0

    work_stack: deque[tuple[int, TreeNode]] = deque([(1, root)])
    max_depth = 1
    while work_stack:
      depth, curr = work_stack.pop()
      max_depth = max(max_depth, depth)
      if curr.left is not None:
        work_stack.append((depth + 1, curr.left))
      if curr.right is not None:
        work_stack.append((depth + 1, curr.right))

    return max_depth
